Documentation/proposals/202305-scrapeclasses.md

```diff
@@ -0,0 +1,223 @@
+## Scrape Classes
+
+* **Owners:**
+  * `nicolastakashi`
+  * `eronwright`
+
+* **Related Tickets:**
+  * https://github.com/prometheus-operator/prometheus-operator/issues/4121
+  * https://github.com/prometheus-operator/prometheus-operator/issues/3922
+  * https://github.com/prometheus-operator/prometheus-operator/issues/5947
+  * https://github.com/prometheus-operator/prometheus-operator/issues/5948
+
+This proposal introduces the concept of *scrape classes*, enabling users to utilize scrape configuration data provided by the administrator, through scrape objects such as PodMonitor, Probe, ServiceMonitor and ScrapeConfig.
+
+## Why
+
+Sometimes Prometheus administrators needs to provide default configurations for scrape objects, such as the definition of TLS certificates, when running Prometheus to scrape pods in an Istio mesh with strict mTLS as described in [Istio documentation](https://istio.io/latest/docs/ops/integrations/prometheus/#tls-settings).
+
+Another motivation is to improve feature parity amongst the monitor resources. The `PodMonitor` and `Probe` resources aren't at parity with `ServiceMonitor` because the latter allows for unsafe TLS settings.
+
+### Pitfalls of the current solution
+
+The only known solution for use cases where you'd need to use unsafe TLS settings is to use `additionalScrapeConfig`.
+The downside is obviously the loss of a great feature of the Prometheus Operator. The monitor resources make it possible
+to compose the scrape configurations in a Kubernetes way.
+
+## Goals
+
+- Allow for the administrator to define a named, reusable scrape configuration snippet, including unsafe elements such as file references.
+- Allow for a user to select a configuration snippet by name in their probe/podmonitor/servicemonitor endpoint configuration.
+- Avoid giving the user the ability to use arbirary files within the Prometheus pod.
+
+## Non-Goals
+
+- Provide a way for Prometheus administrators to enforce/override settings on scrape configurations.
+- Deprecation of the unsafe TLS settings in `ServiceMonitor`.
+
+### Audience
+
+- Users who serve Prometheus as a service and want to give their customers autonomy in defining monitors, but want to provide a default configuration for scraping.
+
+## How
+
+The proposed solution is to introduce a notion of a *scrape class*, akin to a Kubernetes [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/).
+A scrape class defines TLS settings (and possibly other settings in future, e.g. sensitive authorization settings) to be applied to all scrape configs of that class.
+
+One scrape class may be designated as the default class, in which case that class is applied to any scrape config that doesn't specify a value for `scrapeClass`.
+
+When defining a podmonitor/servicemonitor/probe/scrapeconfig, a user may assign a scrape class via the `scrapeClass` field.
+When there's a match, a scrape class is assigned to all the endpoints.
```

Comment:
What then happens if a scrape class is removed? Do we remove the Monitor resources?

Reply:
@xiu yes, this should be the case.

Reply:
I made a couple of tests with Ingress and Ingress Class since the proposal has been changed to work close to this concept.

I tested using NGINX Ingress Controller because it's what I'm most familiar with and I noticed the following behavior.

> What happens when scrapeClassName doesn't match a scrape class of the Prometheus object? The explicit option from my POV is to reject the monitoring object.

When I tried to create an Ingress object referencing an ingressClass that didn't exist I was able to create the manifest but I saw error logs on the ingress pods.

```
I1129 08:56:05.747184       7 store.go:403] "Ignoring ingress because of error while validating ingress class" ingress=" default/ingress-with-auth" error=" no object matching key \"nginx\" in local store"
W1129 08:56:08.715138       7 controller.go:331] ignoring ingress ingress-with-auth in default based on annotation: no object matching key "nginx" in local store
I1129 08:56:08.715190       7 main.go:107] "successfully validated configuration, accepting" ingress="default/ingress-with-auth"
I1129 08:56:08.719779       7 store.go:436] "Ignoring ingress because of error while validating ingress class" ingress=" default/ingress-with-auth" error=" no object matching key \"nginx\" in local store"
```

I don't think this is the Best UX and I think we should reject the object is the class didn't exist, probable through admission webhook to provide a better experience. 

> What then happens if a scrape class is removed? Do we remove the Monitor resources?
I was able to delete the ingress class but when the controller reconciled the ingress object I got the same error as above.

Maybe we could add some finalizers to the scrape class to avoid this issue, tbh I'm not sure about the best strategy here.

Reply:
> Maybe we could add some finalizers to the scrape class

Since the scrape class definition is part of the Prometheus CR, I'm not sure how finalizers would help.

To be clear, I'm fine with the operator rejecting a monitoring object that references a non-existing scrape class name. We have https://github.com/prometheus-operator/prometheus-operator/issues/4481 for implementing admission webhook for monitoring resources + https://github.com/prometheus-operator/prometheus-operator/pull/6030 for emitting Kubernetes events on rejection.

Reply:
>  I'm more in favor of dedicated CR tbh looking forward it will be easier to maintain.

Reading the proposal, I assumed that the ScrapeClass CRD wasn't the chosen option. I'm not convinced that a ScrapeClass CRD is simpler than defining the scrape classes inside the Prometheus resource.

Reply:
At the moment I copied the proposal as is from the original PR.
I have no doubts of having Scrape Classe at Prometheus Level is easier to implement and I do agree it provide a better DX while handling volumes for example.

But in the other hand it's limiting if you want to leverage same scrape classe between Agent and Server and also it may increase Prometheus Object a lot, especially after features such as default relabel configs be added to the Scrape Class.

Another thing to consider is having a Scrape Class CRD makes the Operator more close to the Ingress implementation which is something well know by the  k8s users.

Reply:
Speaking of similarity with IngressClass, should a ScrapeClass CRD be namespace or cluster scoped?

* if namespace scoped, reusing a scrape class for different Prometheus/PrometheusAgent objects becomes limited. 
* if cluster scoped, a ScrapeClass is probably going to match a small number of Prometheus/PrometheusAgent object, due to the volume mounts.

Reply:
Yes!
I see your point the volume mounts couples the scrape class to the prometheus objects indeed and this makes a bit difficult to reuse cross prometheus instances. The solution I see for this is to have volumes definitions on the scrape class, but I'm not convinced about this one.

Ofc we can move forward with the ScrapeClass being defined on the Prometheus object and in the further if we think it makes more sense move to a dedicated CRD we can plan a work towards the new requirements/design and update this doc.

WDYT?

Reply:
Cool 😄 
So just to summarise, let's keep the behaviour we described above.

- The Operator will not allow multiple default scrape classes
- If the endpoint selects a scrape class that didn't exists the Operator will reject the resource.
- If the delete the scrape class in use the Operator will fail the prometheus reconcile.

RIght?

Reply:
Sounds correct!

Reply:
> If the delete the scrape class in use the Operator will fail the prometheus reconcile.

IMHO it should behave the same as if the service monitor specifies a scrape class that doesn't exist

1. The Prometheus object is created with a scrape class named `foo`.
1. Service monitor A is created with `scrapeClassName: foo` -> the service monitor is selected by the Prometheus controller.
2. The Prometheus object is updated resulting in the scrape class `foo` being removed or renamed.
3. Service monitor A is rejected on the next reconciliation.

Reply:
Hmmmm, sounds like a difficult thing to debug in case that happens.

We probably want to make sure we have logs/metrics/events that allow troubleshooting, also docs!

Reply:
I agree but it's not very different from a ServiceMonitor referring to a non-existing secret key for bearer token. #3611 and #3335 should help.

Reply:
That's make total sense!
I've added this information to the doc cc @simonpasquier @ArthurSens